### PR TITLE
Add text for Final Messages to support half-close

### PR DESCRIPTION
**Pull request by tfpauly**
Satisfies #53

### CODE REVIEW

File: draft-ietf-taps-interface.md [hunk 1]
```diff
@@ -978,6 +978,21 @@ Send Action. It is used to mark data safe for certain 0-RTT establishment
...
+closed in the sending direction. For example, TCP-based Connections can
+send a FIN once a Message marked as Final has been completely sent. 
+Protocols that do not support signalling the end of a Connection in a given
+direction will ignore this property.
+
+Note that a final Message must always be sorted to the end of a list of Messages.
+The final property overrides Niceness and any other property that would re-order
+Messages. If another Message is sent after a Message marked as final has already
+been sent on a Connection, the new Message will report an error.
+
```
Comment (mwelzl) [inline draft-ietf-taps-interface.md]: Capitalization of "Final" vs. "final" is inconsistent above - I suggest to write "The Final property overrides..." and "...marked as Final ..."  in the last two sentences above.

File: draft-ietf-taps-interface.md [hunk 2]
```diff
@@ -1130,6 +1145,21 @@ received that reception has failed. Such conditions that irrevocably lead the
 the termination of the Connection are signaled using ConnectionError instead
 (see {{termination}}).
 
+## Receiving Final Messages
+
+The application may check a property to determine if a received Message is
+the final Message on a Connection. For any Message that is marked as final,
```
Comment (mwelzl) [inline draft-ietf-taps-interface.md]: Suggest "marked as Final"

File: draft-ietf-taps-interface.md [hunk 3]
```diff
@@ -1130,6 +1145,21 @@ received that reception has failed. Such conditions that irrevocably lead the
...
+the application can assume that there will be no more Messages received on the
+Connection once the Message has been completely delivered. This corresponds
+to the Final property that may be marked on a sent Message {{send-final}}.
+
+Final Messages are delivered by protocols that support closing one side of a Connection.
+For example, if a Connection using TCP receives a FIN, the application can determine
+that the stream is complete once the final Message has been marked complete.
+Protocols that do not support this functionality, such as UDP, may never deliver a Message
+marked as final.
+
```
[the remark this answers is not in the document]
Comment (britram) [reply draft-ietf-taps-interface.md]: I also don't think we need to go into this level of detail *here*. You might get a Final message, but you, the application, should not rely on it. (Indeed, if the BSD API for TCP had this functionality, any peer that did the send-last-drop-state-start-RSTing-ACKS trick some CDNs use (used?) as a performance optimization would never send Final messages, even though the underlying transport supports them).

First para is okay. Suggest replacing second para with something like "Some transport protocols and peers may not support signaling of a Final message. Applications therefore must not rely on receiving a message marked Final to know that the other endpoint is done sending on a connection"

**Comment by tfpauly** [comment]
Okay, addressed capitalization and took Brian's suggested paragraph (with one tiny change).

**Comment by philsbln** [comment]
Fine with me - go ahead

**Comment by tfpauly** [comment]
@mwelzl, @britram any further comments on this one?

**Comment by britram** [comment]
ship it

**Comment by mwelzl** [comment]
Sorry, seems I should have clicked "approve" earlier. Still being clumsy in github  :(

**Pull request merged**